presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java

```diff
@@ -113,6 +149,44 @@ public PlanNode visitTableScan(TableScanNode node, RewriteContext<Void> context)
                     originalConstraint);
         }
 
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        {
+            PlanNode left = context.rewrite(node.getLeft());
+            PlanNode right = context.rewrite(node.getRight());
+            return new JoinNode(
+                    node.getId(),
+                    node.getType(),
+                    left,
+                    right,
+                    node.getCriteria(),
+                    node.getOutputSymbols(),
+                    node.getFilter().map(this::desugar),
+                    node.getLeftHashSymbol(),
+                    node.getRightHashSymbol(),
+                    node.getDistributionType());
+        }
+
+        @Override
+        public PlanNode visitValues(ValuesNode node, RewriteContext<Void> context)
+        {
+            return new ValuesNode(node.getId(), node.getOutputSymbols(),
+                    node.getRows().stream()
+                            .map(row -> row.stream()
```

Comment:
I do not like nested streams. Can you extract nested stream chain to separate method?